Assets/Source/Components/Actor/ActorDash.cs

```diff
@@ -0,0 +1,157 @@
+        }
+
+        private void SetupDashTimer()
+        {
+            dashTimer = gameObject.AddComponent<IntervalTimerComponent>();
+            dashTimer.UpdateInterval(CooldownTime);
+            dashTimer.AutoReset = false;
+            dashTimer.OnIntervalReached.AddListener(OnIntervalReached);
+        }
+
+        private void OnIntervalReached()
+        {
+            CanDash = true;
+        }
+
+        /// <summary>
+        /// Because the ActorDash can't assign external velocity the actor needs to call this method when trying to dash.
+        /// This determines if a Dash if available and if it is then the dash component is setup and ready.
+        /// </summary>
+        /// <returns>True if the actor has a dash available and is not currently dashing</returns>
+        public bool TrySetupDashLeft()
+        {
+            if (CanDash && !IsDashing)
+            {
+                dashEndLocation = transform.position.x - DashDistance;
+                IsDashing = !IsDashing;
+                dashDirection = DashDirections.Left;
+            }
+
+            return CanDash;
+        }
+
+        /// <summary>
+        /// Because the ActorDash can't assign external velocity the actor needs to call this method when trying to dash.
+        /// This determines if a Dash if available and if it is then the dash component is setup and ready.
+        /// </summary>
+        /// <returns>True if the actor has a dash available and is not currently dashing</returns>
+        public bool TrySetupDashRight()
+        {
+            if (CanDash && !IsDashing)
+            {
+                dashEndLocation = transform.position.x + DashDistance;
+                IsDashing = !IsDashing;
+                dashDirection = DashDirections.Right;
+            }
+
+            return CanDash;
+        }
+
+        /// <summary>
+        /// Call for all frames where the actor should be dashing by checking the value of <see cref="IsDashing"/>
+        /// </summary>
+        /// <param name="externalVelocity">The current external velocity of the actor</param>
+        /// <returns>The calculated velocity that should be assigned to the actor</returns>
+        public Vector2 Dash(Vector2 externalVelocity)
+        {
+            if (CanDash)
+            {
+                if (dashDirection == DashDirections.Left)
+                {
+                    return DashLeft(externalVelocity);
+                }
+                else
+                {
+                    return DashRight(externalVelocity);
+                }
+            }
+
+            return new Vector2(0, 0);
+        }
+
+        private Vector2 DashLeft(Vector2 externalVelocity)
+        {
+            if (transform.position.x < dashEndLocation)
+            {
+                return EndDash(externalVelocity);
+            }
+            else
+            {
+                float xVel = -DashDistance + externalVelocity.x;
+                if (actorRestrictor.DidHitBorder(xVel))
```

Comment:
i dont think this should be done this way.  as long as we're adding to the velocity the actor restrictor should be doing all the restricting itself.

Reply:
it doesn't work so we might have to look at either dashing in a different way or fixing the restrictor. Basically if you dashed while relatively close to the border the velocity was so high that the actor had already moved out of bounds before the lateupdate calculates it would be out of bounds and then the player would get stuck.